iPad: Notifications splitview

frosty opened the pull request:
This PR introduces a split view to Notifications, following on from @kurzee's readability work. Hopefully everything should work as expected. The left pane always shows the notifications list, and the right pane shows the detail for the selected notification.

Things of note:

* I opted to disable full screen for reader posts in Notifications. Firstly because it just felt weird to me to use. I primarily find notifications to be something you skim through, rather than spend time in reading posts. Secondly, because the navigation is somewhat different to Reader and we'd need to do some design work to figure out the best way to integrate it: in Reader, you're always pushing a fullscreen view onto the detail stack, whereas in notifications the fullscreen view would be set as the root of the detail stack. We could always revisit this in the future.
* If the user has no notifications at all (you can test this with a brand new account), we're showing a full screen "No notifications" screen, much like you get if you have no blogs at all.

To test:

* Ensure you test on different devices, in various multitasking states. 
* Ensure that if you move from the split view being collapsed to expanded, that a notification gets selected.
* Ensure that new accounts both a) show a correct 'no notifications' screen, and b) transition correctly to show notifications once one is received.
* Check state restoration – both the selected segment index and the selected notification should be restored. Also check what happens if a new notification arrives whilst the app isn't running.
* Check that the navigation buttons in the reader detail screens work correctly, and that if you transition back to an expanded split view that the correct notification is highlighted.
* Ensure that full screen doesn't activate at any point!

Needs review: @kurzee

kurzee commented:
Found an interesting case when receiving notifications while on the Notifications tab.

1. Be on the Notifications tab, iPad. (probably iPhone too)
2. Select one of the notifications further down the list, keep the selection on screen.
3. On another account, like a post that belongs to the account your logged in to, in app.
4. When the like notification is received the tableView jumps, animated.
5. Now unlike the post you just liked.
6. Notice the cell selection disappears.
7. Like the post again, the cell selection reappears.
8. Unlike the post.
9. Scroll down, notice a completely different cell is now selected.

This could totally be an existing issue. So we can copy-paste to another issue if needed or unrelated to the cell selection handling.

![notifications-split-view-selection](https://cloud.githubusercontent.com/assets/1873422/23010902/8ca000f4-f3e3-11e6-9578-5e31d51a163a.gif)

kurzee commented:
Not sure if this one is related, probably for the `ReaderPostDetailViewController`.

1. In portrait orientation.
2. Select a notification that will lead to a user's site.
3. Select the header or user cell to load that user's site.
4. The posts should load, select a post.
5. Background the app to preserve state, kill the app.
6. Rotate the device or simulator to landscape.
7. Relaunch the app.
8. If the state restores the reader detail post, notice the narrow content that restored as if it was in portrait orientation.

Can definitely create a separate issue here if unrelated.

![simulator screen shot feb 16 2017 12 55 13 am](https://cloud.githubusercontent.com/assets/1873422/23011021/1e33f0e8-f3e4-11e6-8a6b-dea23c13a7f5.png)

cc @aerych

kurzee commented:
This may be a lost cause:

1. On Notifications tab, select Unread segment.
2. Select a notification down the list.
3. Background the app to preserve state, kill the app.
4. Relaunch the app.
5. Notice the loss of the unread cell, probably expected.
6. Also notice the state restoration of the detail view, probably unexpected.

kurzee commented:
This may be expected:

1. Select a notification down the list.
2. Background the app to preserve state, kill the app.
3. Open another app, bring in a multitasking split with WP.
4. Notice the restored state restores Notifications, but not the selection.

kurzee commented:
Looks like we lost cell selection for iPhone when coming back to the root `NotificationsViewController`. It's in the internal build, probably just needs `clearsSelectionOnViewWillAppear`.

kurzee commented:
On iPhone Plus:

1. Launch, hit Notifications.
2. Don't select anything.
3. Rotate to landscape.
4. Notice the detail view has loaded the first notifications, but the cell is not selected.

kurzee commented:
On iPhone Plus:

1. Select a notification down the list.
2. Rotate to landscape.
3. Background the app to preserve state, kill the app.
4. Launch the app, still in landscape, state should restore.
5. Rotate to portrait, notice we lose the detail view and the cell selection.
6. Rotate to landscape.
7. Notice the cell selection is back, but the detail view is loaded with the first notification in the list.

kurzee commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -135,6 +151,59 @@ class NotificationsViewController: UITableViewController {
+        super.decodeRestorableState(with: coder)
+    }
+
+    fileprivate func decodeSelectedNotificationWithCoder(coder: NSCoder) {
```
Can probably make this a bit more Swift-3-y. `decodeSelectedNotification(coder:`.

kurzee commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -135,6 +151,59 @@ class NotificationsViewController: UITableViewController {
+        }
+    }
+
+    fileprivate func decodeSelectedSegmentIndexWithCoder(coder: NSCoder) {
```
Swift-3-y too.

kurzee commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -615,21 +714,35 @@ extension NotificationsViewController {
+                selectRowForNotification(firstNotification, animated: false, scrollPosition: .none)
+                self.tableView(tableView, didSelectRowAt: indexPath)
+            } else {
+                showDetailViewController(UIViewController(), sender: nil)
```
Maybe a comment here for why we're using an empty view controller.

aerych commented:
> Can definitely create a separate issue here if unrelated

If that can be reproduced in the reader as well (guessing it probably can) then yah, a separate issue for sure.

frosty commented:
Thanks for the review, @kurzee! I think I've addressed your comments:

* Unread cell state restoration: if we're saving state and the user has Unread selected, I'm instead saving the state as All. Because they've viewed the notification, it would never be in 'unread' when the app is relaunched, so by switching to All we can keep the same notification selected. [commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/6681/commits/03d9b99c44b1a87062dfdd749b7c6144c11d2369)
* iPhone Plus initial selection in landscape: should be fixed.
* iPhone Plus restoring in landscape then rotating to portrait: should be fixed.
* Cell deselection: should now be animated when compact.
* I can't find an easy fix for [this one](https://github.com/wordpress-mobile/WordPress-iOS/pull/6681#issuecomment-280250980) at the moment, but I don't think it should block us.
* The Reader detail content size issue looks to be separate, as I've managed to recreate it in the Reader.

Care to take another look?

kurzee commented:
>Unread cell state restoration: if we're saving state and the user has Unread selected, I'm instead saving the state as All. Because they've viewed the notification, it would never be in 'unread' when the app is relaunched, so by switching to All we can keep the same notification selected.

Very nice!

>I can't find an easy fix for this one at the moment, but I don't think it should block us.

I'd say that one can remain as is and feels expected.

>Found an interesting case when receiving notifications while on the Notifications tab.

For that one, https://github.com/wordpress-mobile/WordPress-iOS/pull/6681#issuecomment-280248911, it is actually present in `develop`, but a bit more noticeable with the split-view. Can be resolved separately.

kurzee commented:
@frosty everything else looks great! :shipit: unless you want to make some more tweaks

frosty commented:
Sorry @kurzee, I missed https://github.com/wordpress-mobile/WordPress-iOS/pull/6681#issuecomment-280248911. I'm not entirely sure why the row was being deselected (I think it's something that `WPTableViewHandler` is doing), but I've added some code to `tableViewDidChangeContent(_:)` to ensure the row is reselected.

kurzee commented:
>I'm not entirely sure why the row was being deselected (I think it's something that WPTableViewHandler is doing), but I've added some code to tableViewDidChangeContent(_:) to ensure the row is reselected.

The latest fixed the deselection. We do still have the bouncing issue when receiving a notification, but that is definitely an existing issue.

:shipit: 🔥

frosty commented:
Thanks @kurzee! :tada:
